File: framec_tests/cpp/transition_params/transition_params.cpp
// emitted from framec_v0.10.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

#include <unordered_map>
#include <vector>
#include <any>
#include <string>
#include <iostream>

#include "../FrameLang/FrameLang.h"
using namespace std;

//=============== Compartment ==============//

class TransitParamsCompartment
{
public:
    int state;
    
    TransitParamsCompartment(int state)
    {
        this->state = state;
    }
    
    std::unordered_map<std::string, std::any> stateArgs;
    std::unordered_map<std::string, std::any> stateVars;
    std::unordered_map<std::string, std::any> enterArgs;
    std::unordered_map<std::string, std::any> exitArgs;
    FrameEvent *_forwardEvent = nullptr;
};

class TransitParams
{
private:
    TransitParamsCompartment *_compartment_;
    TransitParamsCompartment *_nextCompartment_;
    
    
    
public:
    TransitParams()
    {
        // Create and intialize start state compartment.
        _state_ = static_cast<int>(TransitParamsState::INIT);
        
        _compartment_ = new TransitParamsCompartment(this->_state_);
        _nextCompartment_ = nullptr;
        
        
        // Send system start event
        FrameEvent frame_event(">", std::unordered_map<std::string, std::any>());
        _mux_(&frame_event);
    
    }
    
    // states enum
private:
    enum class TransitParamsState
    {
        INIT = 0,
        A = 1,
        B = 2
    };
    
    //====================== Multiplexer ====================//

private:
    void _mux_(FrameEvent *e)
    {
        if(this->_compartment_->state == static_cast<int>(TransitParamsState::INIT))
        {
            this->_sInit_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(TransitParamsState::A))
        {
            this->_sA_(e);
        }
        else if(this->_compartment_->state == static_cast<int>(TransitParamsState::B))
        {
            this->_sB_(e);
        }
        
        
        if(this->_nextCompartment_ != nullptr)
        {
            TransitParamsCompartment *nextCompartment = this->_nextCompartment_;
            this->_nextCompartment_ = nullptr;
            if (nextCompartment->_forwardEvent != nullptr && 
                nextCompartment->_forwardEvent->_message == ">")
            {
                this->_mux_(new FrameEvent( "<", this->_compartment_->exitArgs));
                this->_compartment_ = nextCompartment;
                this->_mux_(nextCompartment->_forwardEvent);
            }
            else
            {
                this->_doTransition_(nextCompartment);
                if (nextCompartment->_forwardEvent != nullptr)
                {
                    this->_mux_(nextCompartment->_forwardEvent);
                }
            }
            nextCompartment->_forwardEvent = nullptr;
        }
    }
    
        
        //===================== Interface Block ===================//
    public:
        void Next() {
            FrameEvent e("Next", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
        void Change() {
            FrameEvent e("Change", unordered_map<string, any>());
            this->_mux_(&e);
        }
        
    
    //===================== Machine Block ===================//

private:
    
    void _sInit_(FrameEvent *e)
    {
        if (e->_message == "Next") {
            TransitParamsCompartment *compartment =  new TransitParamsCompartment(static_cast<int>(TransitParamsState::A));
            compartment->enterArgs["msg"] = std::string("hi A");
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Change") {
            TransitParamsCompartment *compartment =  new TransitParamsCompartment(static_cast<int>(TransitParamsState::A));
            
            this->_changeState_(compartment);
            return;
        }
    }
    
    void _sA_(FrameEvent *e)
    {
        if (e->_message == ">") {
            log_do(any_cast<string>(e->_parameters["msg"]));
            return;
        }
        else if (e->_message == "<") {
            log_do("bye A");
            return;
        }
        else if (e->_message == "Next") {
            TransitParamsCompartment *compartment =  new TransitParamsCompartment(static_cast<int>(TransitParamsState::B));
            compartment->enterArgs["msg"] = std::string("hi B");
            compartment->enterArgs["val"] = 42;
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Change") {
            TransitParamsCompartment *compartment =  new TransitParamsCompartment(static_cast<int>(TransitParamsState::B));
            
            this->_changeState_(compartment);
            return;
        }
    }
    
    void _sB_(FrameEvent *e)
    {
        if (e->_message == ">") {
            log_do(any_cast<string>(e->_parameters["msg"]));
            log_do(to_string(e->_parameters["val"]));
            return;
        }
        else if (e->_message == "<") {
            log_do(to_string(e->_parameters["val"]));
            log_do(any_cast<string>(e->_parameters["msg"]));
            return;
        }
        else if (e->_message == "Next") {
            this->_compartment_->exitArgs["val"] = true;
            this->_compartment_->exitArgs["msg"] = std::string("bye B");
            TransitParamsCompartment *compartment =  new TransitParamsCompartment(static_cast<int>(TransitParamsState::A));
            compartment->enterArgs["msg"] = std::string("hi again A");
            
            this->_transition_(compartment);
            return;
        }
        else if (e->_message == "Change") {
            TransitParamsCompartment *compartment =  new TransitParamsCompartment(static_cast<int>(TransitParamsState::A));
            
            this->_changeState_(compartment);
            return;
        }
    }

//===================== Actions Block ===================//
    
    
    
public:
    void log_do(const std::string& msg)
    {
        tape.push_back(msg);
        for (const auto& msg : tape)
        {
            std::cout << msg << std::endl;
        }
    }
    
    // Unimplemented Actions
    public:
    
    //===================== Domain Block ===================//
    
    
    public:
    std::vector<std::string> tape ;
    
    
    //=============== Machinery and Mechanisms ==============//

private:
    int _state_;
    
    void _transition_(TransitParamsCompartment *compartment)
    {
        _nextCompartment_ = compartment;
    }
    
    void _doTransition_(TransitParamsCompartment *nextCompartment)
    {
        this->_mux_(new FrameEvent("<", _compartment_->exitArgs));
        this->_compartment_ = nextCompartment;
        this->_mux_(new FrameEvent(">", this->_compartment_->enterArgs));
    }
    
    private: void _changeState_(TransitParamsCompartment* compartment)
    {
        this->_compartment_ = compartment;
    }
    
    public:
    string state_info(){
        return std::to_string(_compartment_->state);
        }
        
};

/********************

class TransitParamsController : public TransitParams
{
public:
	TransitParamsController() : TransitParams() {}
};

********************/

File: framec_tests/javascript/src/transition_params/transition_params.js
// emitted from framec_v0.11.0
// get include files at https://github.com/frame-lang/frame-ancillary-files

function FrameEvent(message, parameters) {

    var that = {};
    that._message = message;
    that._parameters = parameters;
    that._return = null;
    return that;
    
}

class TransitParams {
    
    // creating private properties
    
    #state
    #compartment
    #nextCompartment
    
    constructor () {
        
        // Create and intialize start state compartment.
        
        this.#state = this.#sInit_;
        this.#compartment = new TransitParamsCompartment(this.#state);
        this.#nextCompartment = null;
        
        // Initialize domain
        this.tape = [];
        
        // Send system start event
        const frameEvent = FrameEvent(">", null);
        this.#mux(frameEvent);
    }
    
    //===================== Interface Block ===================//
    
    Next() {
        let e = FrameEvent("Next",null);
        this.#mux(e);
    }
    
    Change() {
        let e = FrameEvent("Change",null);
        this.#mux(e);
    }
    
    //====================== Multiplexer ====================//
    
    #mux(e) {
        switch (this.#compartment.state) {
            case this.#sInit_:
                this.#sInit_(e);
                break;
            case this.#sA_:
                this.#sA_(e);
                break;
            case this.#sB_:
                this.#sB_(e);
                break;
        }
        
        if( this.#nextCompartment != null) {
            let nextCompartment = this.#nextCompartment
            this.#nextCompartment = null
            if (nextCompartment._forwardEvent != null && 
               nextCompartment._forwardEvent._message == ">") {
                this.#mux(FrameEvent( "<", this.#compartment.ExitArgs))
                this.#compartment = nextCompartment
                this.#mux(nextCompartment._forwardEvent)
            } else {
                this.#doTransition(nextCompartment)
                if (nextCompartment._forwardEvent != null) {
                    this.#mux(nextCompartment._forwardEvent)
                }
            }
            nextCompartment._forwardEvent = null
        }
    }
    
    //===================== Machine Block ===================//
    
    #sInit_(e) {
        switch (e._message) {
            case "Next":
                {
                let compartment =  new TransitParamsCompartment(this.#sA_);
                
                compartment.EnterArgs["msg"] = "hi A";
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Change":
                {
                
                return;
                }
                
        }
    }  // ->> $A
	
    
    #sA_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do((e._parameters["msg"]));
                
                return;
                }
                
            case "<":
                {
                this.log_do("bye A");
                
                return;
                }
                
            case "Next":
                {
                let compartment =  new TransitParamsCompartment(this.#sB_);
                
                compartment.EnterArgs["msg"] = "hi B";
                compartment.EnterArgs["val"] = 42;
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Change":
                {
                
                return;
                }
                
        }
    }  // ->> $B
	
    
    #sB_(e) {
        switch (e._message) {
            case ">":
                {
                this.log_do((e._parameters["msg"]));
                this.log_do((e._parameters["val"]).toString());
                
                return;
                }
                
            case "<":
                {
                this.log_do((e._parameters["val"]).toString());
                this.log_do((e._parameters["msg"]));
                
                return;
                }
                
            case "Next":
                {
                this.#compartment.ExitArgs["val"] = true;
                this.#compartment.ExitArgs["msg"] = "bye B";
                let compartment =  new TransitParamsCompartment(this.#sA_);
                
                compartment.EnterArgs["msg"] = "hi again A";
                
                this.#transition(compartment);
                
                return;
                }
                
            case "Change":
                {
                
                return;
                }
                
        }
    }
    
    //===================== Actions Block ===================//
    
    // Unimplemented Actions
    
    log_do(msg) { throw new Error('Action not implemented.'); }
    
    //=============== Machinery and Mechanisms ==============//
    
    #transition(compartment) {
        this.#nextCompartment = compartment;
    }
    
    #doTransition(nextCompartment) {
        this.#mux(FrameEvent("<", this.#compartment.ExitArgs));
        this.#compartment = nextCompartment;
        this.#mux(FrameEvent(">", this.#compartment.EnterArgs));
    }
    
    state_info() {
        return this.#compartment.state.name;
    }
    
      // ->> $A
	
    
};

//=============== Compartment ==============//

class TransitParamsCompartment {

    constructor(state) {
        this.state = state
    }
    
    StateArgs = {};
    StateVars = {};
    EnterArgs = {};
    ExitArgs = {};
    _forwardEvent = FrameEvent.call(this)
}


/********************

class TransitParamsController extends TransitParams {

	constructor() {
	  super()
	}
	log_do(msg) {}
};

********************/

module.exports = TransitParams
